src/sql/workbench/parts/notebook/common/models/notebookTextFileModel.ts

```diff
@@ -0,0 +1,256 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Range, IRange } from 'vs/editor/common/core/range';
+import { UntitledEditorModel } from 'vs/workbench/common/editor/untitledEditorModel';
+import { TextFileEditorModel } from 'vs/workbench/services/textfile/common/textFileEditorModel';
+import { FindMatch } from 'vs/editor/common/model';
+import { NotebookContentChange, INotebookModel } from 'sql/workbench/parts/notebook/common/models/modelInterfaces';
+import { NotebookChangeType } from 'sql/workbench/parts/notebook/common/models/contracts';
+
+export class NotebookTextFileModel {
+	// save active cell's line/column in editor model for the beginning of the source property
+	private sourceBeginRange: Range;
+	// save active cell's line/column in editor model for the beginning of the output property
+	private outputBeginRange: Range;
+	// save active cell guid
+	private activeCellGuid: string;
+
+	constructor() {
+	}
+
+	public get ActiveCellGuid(): string {
+		return this.activeCellGuid;
+	}
+
+	public set ActiveCellGuid(guid: string) {
+		if (this.activeCellGuid !== guid) {
+			this.sourceBeginRange = undefined;
+			this.outputBeginRange = undefined;
+			this.activeCellGuid = guid;
+		}
+	}
+
+	public transformAndApplyEditForSourceUpdate(contentChange: NotebookContentChange, textEditorModel: TextFileEditorModel | UntitledEditorModel): boolean {
+		let cellGuidRange = this.getCellNodeByGuid(textEditorModel, contentChange.cells[0].cellGuid);
+
+		// convert the range to leverage offsets in the json
+		if (contentChange && contentChange.modelContentChangedEvent && this.areRangePropertiesPopulated(cellGuidRange)) {
+			contentChange.modelContentChangedEvent.changes.forEach(change => {
+				let convertedRange: IRange = {
+					startLineNumber: change.range.startLineNumber + cellGuidRange.startLineNumber - 1,
+					endLineNumber: change.range.endLineNumber + cellGuidRange.startLineNumber - 1,
+					startColumn: change.range.startColumn + cellGuidRange.startColumn,
+					endColumn: change.range.endColumn + cellGuidRange.startColumn
+				};
+				// Need to subtract one because we're going from 1-based to 0-based
+				let startSpaces: string = ' '.repeat(cellGuidRange.startColumn - 1);
+				// The text here transforms a string from 'This is a string\n this is another string' to:
+				//     This is a string
+				//     this is another string
+				textEditorModel.textEditorModel.applyEdits([{
+					range: new Range(convertedRange.startLineNumber, convertedRange.startColumn, convertedRange.endLineNumber, convertedRange.endColumn),
+					text: change.text.split('\n').join('\\n\",\n'.concat(startSpaces).concat('\"'))
```

Comment:
>\n [](start = 30, length = 2)

This functions correctly on a windows machine too?

Reply:
The \r is still there and we're putting the \n back in afterwards. In my testing, it functioned as expected.